Marlin/src/HAL/STM32/Sd2Card_sdio_stm32duino.cpp

```diff
@@ -88,18 +88,37 @@
 
     MKS Robin board seems to have stable SDIO with BusWide 1bit and ClockDiv 8 i.e. 4.8MHz SDIO clock frequency
     Additional testing is required as there are clearly some 4bit initialization problems
-
-    Add -DTRANSFER_CLOCK_DIV=8 to build parameters to improve SDIO stability
   */
 
-  #ifndef TRANSFER_CLOCK_DIV
-    #define TRANSFER_CLOCK_DIV (uint8_t(SDIO_INIT_CLK_DIV) / 40)
-  #endif
-
   #ifndef USBD_OK
     #define USBD_OK 0
   #endif
 
+  // Target Clock, configurable. Default is 18MHz, from STM32F1
+  #ifndef SDIO_CLOCK
+    #define SDIO_CLOCK                         18000000       /* 18 MHz */
+  #endif
+
+  // SDIO retries, configurable. Default is 3, from STM32F1
+  #ifndef SDIO_READ_RETRIES
+    #define SDIO_READ_RETRIES                  3
+  #endif
+
+  // SDIO Max Clock (naming from STM Manual, don't change)
+  #define SDIOCLK 48000000
+
+  static uint32_t sdio_set_clock(uint32_t clk) {
+    /*
+     * limit the SDIO master clock to 8/3 of PCLK2. See STM32 Manuals
+     * Also limited to no more than 48Mhz (SDIOCLK).
+     */
+    const uint32_t pclk2 = HAL_RCC_GetPCLK2Freq();
+    clk = min(clk, ((uint32_t)pclk2/3)*8);
+    clk = min(clk, (uint32_t)SDIOCLK);
+    // round up divider, so we don't run the card over the speed supported.
+    return pclk2/clk + (pclk2 % clk != 0) - 2;
```

Comment:
How is this rounding up?

It looks like it _sometimes_ adds one, but _always_ subtracts 2. Without rounding would this already have a `- 2`?

Reply:
The comment explains why it's round up, in the case of remainder, it's better to get a higher divider to avoid pass the asked speed.

The `-2` it's because stm32duino will add `+ 2` to the divider we pass, as documented in their documentation.

Reply:
`SDIO clock frequency is 48MHz / (TRANSFER_CLOCK_DIV + 2)`

Reply:
Could you add a comment reflecting that? I find it best if you don't have to look up reference manuals for something like this. From the reference manual:

`// SDIO_CK frequency = SDIOCLK / [CLKDIV + 2]`